Make DeviceDomain exit immediately if doctor says it can't list anything

tvolkert opened the pull request:
This behavior matches that of `DeviceCommand`. It will means that the
`flutter daemon` process will immediately exit with an exitCode of 1
if it can't find at least one workflow that reports it's able to list
devices.

tvolkert committed:
Make DeviceDomain exit immediately if doctor says it can't list anything

This behavior matches that of `DeviceCommand`. It will means that the
`flutter daemon` process will immediately exit with an exitCode of 1
if it can't find at least one workflow that reports it's able to list
devices.

tvolkert commented:
Devon & Brian, with this change, in the bug we were tracking today whereby the daemon process couldn't locate any Android SDK... the process will exit rather than staying resident and silently reporting that no devices are connected.

tvolkert committed:
Fix tests

tvolkert committed:
Alphabetically sort imports

devoncarew commented:
Hmm, should we exit the flutter daemon, or add a mechanism to report errors to the user (but keep the process alive)? How will they fix their system? I assume we'll need to fix their system or paths, and then restart IntelliJ to bring the daemon back up. I'm wondering if we should make the daemon self-healing in some sense.

skybrian commented:
Based on the code in DeviceDaemon and DeviceService, we don't currently display any error message if the daemon dies. It just doesn't show any devices. (DeviceService.getStatus() returns INACTIVE.)

It currently starts or restarts the daemon on project open, when the Flutter SDK changes, or when the Bazel workspace changes. The user could force it to try again by closing and reopening the project.

So, this change shouldn't make things any worse.

I think it would be reasonable to show an error message in the device dropdown:

```
    Cannot determine devices <Run Flutter Doctor> <Try Again>
```

skybrian commented on packages/flutter_tools/lib/src/commands/daemon.dart:
```diff
@@ -524,6 +536,9 @@ typedef void _DeviceEventHandler(Device device);
 class DeviceDomain extends Domain {
   DeviceDomain(Daemon daemon) : super(daemon, 'device') {
+    if (!doctor.canListAnything) {
+      daemon.shutdown(error: new ToolExit('Unable to locate a development device', exitCode: 1));
```
Can we say anything more specific here? For example, this would be a great time to show ANDROID_HOME.

tvolkert commented:
What would yield the best UX and cleanest code on both sides (tools and plugin)?  Would it be for the daemon process to die, or for it to emit an error event and stay alive?  I'm thinking the latter (which means this PR requires some updating).

skybrian commented on packages/flutter_tools/test/daemon_test.dart:
```diff
@@ -46,7 +33,7 @@ void main() {
     });
 
-    _testUsingContext('daemon.version', () async {
+    testUsingContext('daemon.version', () async {
```
How about: "daemon.version command should succeed"

skybrian commented on packages/flutter_tools/test/daemon_test.dart:
```diff
@@ -63,7 +50,7 @@ void main() {
     });
 
-    _testUsingContext('daemon.logMessage', () {
+    testUsingContext('daemon.logMessage', () {
```
"printError should send daemon.logMessage event"

skybrian commented on packages/flutter_tools/test/daemon_test.dart:
```diff
@@ -86,7 +73,7 @@ void main() {
     });
 
-    _testUsingContext('daemon.logMessage logToStdout', () async {
+    testUsingContext('daemon.logMessage logToStdout', () async {
```
"printStatus should log to stdout when logToStdout is enabled"

skybrian commented on packages/flutter_tools/test/daemon_test.dart:
```diff
@@ -110,7 +97,7 @@ void main() {
     });
 
-    _testUsingContext('daemon.shutdown', () async {
+    testUsingContext('daemon.shutdown', () async {
```
"daemon.shutdown command should stop daemon"

skybrian commented on packages/flutter_tools/test/daemon_test.dart:
```diff
@@ -126,7 +113,7 @@ void main() {
     });
 
-    _testUsingContext('daemon.start', () async {
+    testUsingContext('daemon.start', () async {
```
"app.start without a deviceId should report an error"

skybrian commented on packages/flutter_tools/test/daemon_test.dart:
```diff
@@ -147,7 +134,7 @@ void main() {
     });
 
-    _testUsingContext('daemon.restart', () async {
+    testUsingContext('daemon.restart', () async {
```
"app.restart without an appId should report an error"

skybrian commented on packages/flutter_tools/test/daemon_test.dart:
```diff
@@ -168,7 +155,7 @@ void main() {
     });
 
-    _testUsingContext('daemon.callServiceExtension', () async {
+    testUsingContext('daemon.callServiceExtension', () async {
```
"ext.flutter.debugPaint via service extension without an appId should report an error"

skybrian commented on packages/flutter_tools/test/daemon_test.dart:
```diff
@@ -195,7 +182,7 @@ void main() {
     });
 
-    _testUsingContext('daemon.stop', () async {
+    testUsingContext('daemon.stop', () async {
```
"app.stop without appId should report an error"

skybrian commented on packages/flutter_tools/test/daemon_test.dart:
```diff
@@ -216,7 +203,20 @@ void main() {
     });
 
-    _testUsingContext('device.getDevices', () async {
+    testUsingContext('device when !doctor.canListAnything', () async {
```
"daemon should fail on startup if no devices are available"

skybrian commented on packages/flutter_tools/test/daemon_test.dart:
```diff
@@ -232,7 +232,7 @@ void main() {
     });
 
-    _testUsingContext('device.notify', () {
+    testUsingContext('device.notify', () {
```
"should send device.added event when device is discovered"

skybrian commented:
Yes, I think sending an error might be better so that no restart is required.

If we're willing to change the wire protocol, from the client's point of view, it would be convenient if we could assume that each device error we receive replaces the previous one, and there is an event to clear the error. (Otherwise it's not clear when to remove the error in the UI.) I guess this would be device.updateStatusMessage or something like that.

skybrian commented:
Similarly, the device.added and device.removed events are less convenient than just sending the whole list again when something changed. It would be most convenient if each device event entirely replaced the previous state, either with a list of devices or an error.

tvolkert commented:
> It would be most convenient if each device event entirely replaced the previous state

Since we poll for devices in tools, that could get awfully chatty.  What about keeping the existing `device.added` and `device.removed`, and adding a new `device.replaceAll` and `device.error`?

Hixie commented on packages/flutter_tools/lib/src/commands/daemon.dart:
```diff
@@ -66,7 +67,12 @@ class DaemonCommand extends FlutterCommand {
-    printError('Error from flutter daemon: $error', stackTrace: stackTrace);
+    if (error is ToolExit) {
+      printError(error.message);
+      exit(error.exitCode);
```
shouldn't this go through the normal logic for application exit?

Hixie replied:
otherwise we won't report crashes, etc.

cc @yjbanov

tvolkert replied:
Yeah, in my local edits, I removed this error handler - not sure why it was there...

skybrian commented:
re: could get awfully chatty

Assuming no event is sent when nothing changed, it's the same number of messages and the message size is bounded (users aren't likely to have many devices) so the compression from sending diffs to the current state doesn't seem worth it. But it's working, so maybe not worth the migration?

Adding `device.replaceAll` and `device.setStatus` seems fine; we could stop sending the other ones later, maybe.

devoncarew commented:
I'm partial to the `add` and `remove` events - it lets clients determine when devices are added and removed w/o diffing lists.

Re: client notification, I was thinking of a new api event, like `daemon.showMessage` or some such. This would be similar to the existing `logMessage` (https://github.com/flutter/flutter/wiki/The-flutter-daemon-mode#logmessage). This would be for displaying toasts in the connected clients; the clients would be responsible for things like how to display the message and time to live, and the daemon would need to ensure to not sent too many.

skybrian commented:
Okay, I was thinking of showing it in the device dropdown instead of a toast, but I guess that would work.

There is a minor glitch with only using add/remove: when we restart the device daemon, we have to clear the device list to avoid duplicates, so the user sees the devices disappear and reappear when nothing really happened. With replaceAll(), we can keep displaying the old device list until the first event shows up from the new daemon. But to fix that we'd only need to send replaceAll() at startup, or otherwise indicate that the initial device list is complete.

tvolkert committed:
Review round 1

tvolkert commented:
Ok, new commit pushed.  It sends a `daemon.showMessage` event rather than dying.  We can add the `device.replaceAll` in the future -- with this change, it's up to IntelliJ to show the toast, and it's then up to the user to run doctor and fix whatever's wrong (requiring https://github.com/flutter/flutter-intellij/issues/985)

skybrian commented:
Sounds good. (Once tests are fixed.)

tvolkert committed:
Fix test

devoncarew commented on packages/flutter_tools/lib/src/commands/daemon.dart:
```diff
@@ -539,6 +542,12 @@ class DeviceDomain extends Domain {
+    if (!discoverer.canListAnything)
+      sendEvent(
+          'daemon.showMessage',
+          'Unable to discover ${discoverer.name}. Please run "flutter doctor" '
```
We may want to include a title and message field here. So, something like:

```
{
  'title': 'Unable to list devices',
  'message': 'Unable to list ${'discoverer.name}; please run "flutter doctor" to diagnose potential issues.'
}
```

or similar

devoncarew replied:
(A client UI will likely have a title and message field, and it'd be nice if the daemon provided a useful title for each message).

tvolkert replied:
Good point - done.

devoncarew commented:
This is awesome! I have a suggestion for a bit more structure in the message to the client, in order to better populate our toast. Happy to hook up the client side of this.

devoncarew commented:
lgtm

tvolkert committed:
Review comment

tvolkert committed:
Fix test
